maildir.5
.TH maildir 5
.SH "NAME"
maildir \- directory for incoming mail messages
.SH "INTRODUCTION"
.I maildir
is a structure for
directories of incoming mail messages.
It solves the reliability problems that plague
.I mbox
files and
.I mh
folders.
.SH "RELIABILITY ISSUES"
A machine may crash while it is delivering a message.
For both
.I mbox
files and
.I mh
folders this means that the message will be silently truncated.
Even worse: for
.I mbox
format, if the message is truncated in the middle of a line,
it will be silently joined to the next message.
The mail transport agent will try again later to deliver the message,
but it is unacceptable that a corrupted message should show up at all.
In
.IR maildir ,
every message is guaranteed complete upon delivery.

A machine may have two programs simultaneously delivering mail
to the same user.
The
.I mbox
and
.I mh
formats require the programs to update a single central file.
If the programs do not use some locking mechanism,
the central file will be corrupted.
There are several
.I mbox
and
.I mh
locking mechanisms,
none of which work portably and reliably.
In contrast, in
.IR maildir ,
no locks are ever necessary.
Different delivery processes never touch the same file.

A user may try to delete messages from his mailbox at the same
moment that the machine delivers a new message.
For
.I mbox
and
.I mh
formats, the user's mail-reading program must know
what locking mechanism the mail-delivery programs use.
In contrast, in
.IR maildir ,
any delivered message
can be safely updated or deleted by a mail-reading program.

Many sites use Sun's 
.B Network F\fPa\fBil\fPur\fBe System
(NFS),
presumably because the operating system vendor does not offer
anything else.
NFS exacerbates all of the above problems.
Some NFS implementations don't provide
.B any
reliable locking mechanism.
With 
.I mbox
and
.I mh
formats,
if two machines deliver mail to the same user,
or if a user reads mail anywhere except the delivery machine,
the user's mail is at risk.
.I maildir
works without trouble over NFS.
.SH "THE MAILDIR STRUCTURE"
A directory in
.I maildir
format has three subdirectories,
all on the same filesystem:
.BR tmp ,
.BR new ,
and
.BR cur .

Each file in
.B new
is a newly delivered mail message.
The modification time of the file is the delivery date of the message.
The message is delivered
.I without
an extra UUCP-style
.B From_
line,
.I without
any
.B >From
quoting,
and
.I without
an extra blank line at the end.
The message is normally in RFC 822 format,
starting with a
.B Return-Path
line and a
.B Delivered-To
line,
but it could contain arbitrary binary data.
It might not even end with a newline.

Files in
.B cur
are just like files in
.BR new .
The big difference is that files in
.B cur
are no longer new mail:
they have been seen by the user's mail-reading program.
.SH "HOW A MESSAGE IS DELIVERED"
The
.B tmp
directory is used to ensure reliable delivery,
as discussed here.

A program delivers a mail message in six steps.
First, it
.B chdir()\fPs
to the
.I maildir
directory.
Second, it 
.B stat()s
the name
.BR tmp/\fItime.pid.host ,
where
.I time
is the number of seconds since the beginning of 1970 GMT,
.I pid
is the program's process ID,
and
.I host
is the host name.
Third, if
.B stat()
returned anything other than ENOENT,
the program sleeps for two seconds, updates
.IR time ,
and tries the
.B stat()
again, a limited number of times.
Fourth, the program
creates
.BR tmp/\fItime.pid.host .
Fifth, the program
.I NFS-writes
the message to the file.
Sixth, the program
.BR link() s
the file to
.BR new/\fItime.pid.host .
At that instant the message has been successfully delivered.

The delivery program is required to start a 24-hour timer before
creating
.BR tmp/\fItime.pid.host ,
and to abort the delivery
if the timer expires.
Upon error, timeout, or normal completion,
the delivery program may attempt to
.B unlink()
.BR tmp/\fItime.pid.host .

.I NFS-writing
means
(1) as usual, checking the number of bytes returned from each
.B write()
call;
(2) calling
.B fsync()
and checking its return value;
(3) calling
.B close()
and checking its return value.
(Standard NFS implementations handle
.B fsync()
incorrectly
but make up for it by abusing
.BR close() .)
.SH "HOW A MESSAGE IS READ"
A mail reader operates as follows.

It looks through the
.B new
directory for new messages.
Say there is a new message,
.BR new/\fIunique .
The reader may freely display the contents of
.BR new/\fIunique ,
delete
.BR new/\fIunique ,
or rename
.B new/\fIunique
as
.BR cur/\fIunique:info .
See
.B https://cr.yp.to/docs/smtplf.html/proto/maildir.html
for the meaning of
.IR info .

The reader is also expected to look through the
.B tmp
directory and to clean up any old files found there.
A file in
.B tmp
may be safely removed if it
has not been accessed in 36 hours.

It is a good idea for readers to skip all filenames in
.B new
and
.B cur
starting with a dot.
Other than this, readers should not attempt to parse filenames.
.SH "ENVIRONMENT VARIABLES"
Mail readers supporting
.I maildir
use the
.B MAILDIR
environment variable
as the name of the user's primary mail directory.
.SH "SEE ALSO"
mbox(5),
qmail-local(8)


CHANGELOG
Changelog

20240926 Roberto Puzzanghera
         - code poliched from several errors/warnings against latest GCC 14.2
         - merged auth-tls-forcetls patch v. 2024.07.10

-------------------------------------------------------------------------------------

20071130 version: netqmail 1.06
20071130 legal: qmail-1.03 is now in the public domain
20051103 doc: dot-qmail.9 updated for changed (19980613) conf-patrn default, tnx ADM
20040121 version: netqmail 1.05
20040121 code: qmail-smtpd is protected from exceedingly long (eg 2GB)
         header lines
20040121 code: qmail_lspawn, qmail-newmrh, qmail-newu, and qmail-rspawn
         are protected from misbehaving on hosts where the size of an
         integer is not the same as the size of a character pointer
         (eg 64 bit hosts with 32 bit ints)
20031027 version: netqmail 1.04.
20031027 doc: INSTALL points to http://lifewithqmail.org/lwq.html
20031027 doc: qmail.7 identifies installation as netqmail and points to
         http://qmail.org/
20031027 doc: qmail-queue.8 adds explanation of $QMAILQUEUE
20031027 doc: qmail-log.5 adds reference to errors from $QMAILQUEUE script
20031027 doc: FAQ also points to http://cr.yp.to/qmail/faq.html and
         http://qmail.org/
20031027 code: qmail-smtpd identifies itself as netqmail
20031027 code: if $QMAILQUEUE is set, it's invoked instead of qmail-queue
20031024 code: changed errno from int to #include.
20031024 code: fixed .qmail parsing bug.
20031024 code: recognize 0.0.0.0 as a local address.
20031024 code: sendmail's -f flag now overrides environment variables.
19980615 version: qmail 1.03.
19980614 doc: eliminated BIN.setup in favor of a web page.
19980614 code: added other auto* to qmail-showctl output.
19980614 doc: added pointer to immhf.html in qmail-header.5.
19980614 doc: added note to TEST.receive about SMTP command format.
19980614 doc: added FAQ 5.6 on qmail-qmqpd.
19980614 code: removed unused variables in idedit.c.
19980613 code: changed conf-patrn to 002.
19980613 doc: moved SENDMAIL lower in INSTALL.
19980612 code: added install-big.
19980612 code: added BIN.Makefile.
19980612 doc: added BIN.README, BIN.setup.
19980612 code: switched to new install.
19980611 code: added idedit.
19980611 doc: added FAQ 1.3 on $QMAILMFTFILE.
19980611 doc: used bouncesaying in FAQ 5.5.
19980611 code: added except.
19980611 code: added bouncesaying.
19980611 code: allowed unbracketed IP addresses in dns_ipplus() and
         dns_mxip().
19980611 code: allowed spaces after colon in non-bracketed addresses in
         qmail-smtpd.
19980611 doc: cleaned up UPGRADE.
19980528 bug: qmail-smtpd skips first character in rcpthosts() call.
         tnx NND. impact: qmail-smtpd crashes on empty address; and it
	 allows relaying to ""@any.host. fix: use addr.s.
19980515 doc: expanded flock discussion in INSTALL.mbox.
19980515 doc: eliminated flock warning from INSTALL.maildir.
19980515 doc: split REMOVE.binmail out of INSTALL.
19980515 doc: split REMOVE.sendmail out of INSTALL.
19980515 doc: split TEST.deliver and TEST.receive out of INSTALL and
         UPGRADE.
19980515 doc: integrated INSTALL.boot into INSTALL.
19980515 code: cleaned up final output in qmail-qmqpd.c.
19980514 doc: updated procmail notes in INSTALL.mbox. tnx JRM.
19980514 doc: changed FAQ 4.4 to point to INSTALL.mbox for procmail.
         tnx JRM.
19980514 code: separated HELO and EHLO; single-line response for HELO.
         tnx to various people.
19980430 version: qmail 1.02.
19980430 doc: updated SECURITY.
19980430 doc: fixed FAQ 4.9. tnx KB.
19980430 code: changed quote2() to avoid quoting <>.
19980429 code: changed quote_need() to quote empty local parts. tnx HHO.
19980428 doc: added status notes to INSTALL and UPGRADE.
19980428 code: skip setting environment in sendmail.c if PROTO is set.
19980428 code: eliminated recipientmap.
19980428 code: added virtual users to qmail-send.c. tnx RN.
19980428 code: eliminated domain from rewrite() in qmail-send.c.
19980428 code: added binm1, binm1+df, binm2, binm2+df, binm3, binm3+df.
19980428 doc: eliminated most Mailbox references from INSTALL, UPGRADE.
19980428 code: added config-fast.
19980428 code: renamed qmail-config as config.
19980428 code: supported QMAILMFTFILE in qmail-inject.c.
19980428 code: recognized Mail-Followup-To in hfield.c.
19980428 code: replaced rwrecip() with rwappend() in qmail-inject.c.
19980428 code: cleaned up doheaderfile() in qmail-inject.c.
19980426 code: eliminated -type test from qmail-qstat to speed it up.
         tnx FT.
19980421 doc: eliminated remove-rcpthosts comments from FAQ.
19980421 doc: updated FAQ 4.3 to point to Russ Allbery's FAQ. 
19980421 doc: took account of /var/qmail/boot in INSTALL, UPGRADE, and
         INSTALL.vsm.
19980421 code: added /var/qmail/boot, with home, home+df, proc, proc+df.
19980421 doc: skipped make and make man in INSTALL.
19980420 doc: cleaned up mbox description in SENDMAIL.
19980420 code: changed QMQP port to official port 628.
19980402 doc: updated qmsmac references to fastforward.
19980402 doc: replaced qmail-upgrade man page with doc/SENDMAIL.
19980402 code: added qmqpservers output to qmail-showctl.
19980402 code: added qmail-qmqpd.
19980402 code: added qmail-qmqpc.
19980304 code: eliminated del_saywhynoexit in qmail-send.c.
19980304 code: eliminated concurrencynodel in qmail-send.c.
19980222 code: added status() to qmail-send.c.
19980222 code: added concurrencyused to qmail-send.c.
19980128 doc: added note to qmail-getpw.9 about ETXTBSY.
19980127 code: eliminated err_seenmail() in qmail-smtpd.c. tnx PO.
19980126 doc: used $DEFAULT in FAQ where possible.
19980126 code: added DEFAULT in qmail-local.
19980126 code: added -/ to qmail-pw2u.
19980126 code: revamped qmeopen() as qmesearch() with more sensible
         semantics, separating dash from ext cleanly.
19980126 code: split qmeexists() out of qmeopen() in qmail-local.c.
19980126 code: introduced safeext in qmail-local.c.

maildir.c
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include "prioq.h"
#include "env.h"
#include "stralloc.h"
#include "direntry.h"
#include "datetime.h"
#include "now.h"
#include "str.h"
#include "maildir.h"

struct strerr maildir_chdir_err;
struct strerr maildir_scan_err;

int maildir_chdir()
{
 char *maildir;
 maildir = env_get("MAILDIR");
 if (!maildir)
   STRERR(-1,maildir_chdir_err,"MAILDIR not set")
 if (chdir(maildir) == -1)
   STRERR_SYS3(-1,maildir_chdir_err,"unable to chdir to ",maildir,": ")
 return 0;
}

void maildir_clean(tmpname)
stralloc *tmpname;
{
 DIR *dir;
 direntry *d;
 datetime_sec time;
 struct stat st;

 time = now();

 dir = opendir("tmp");
 if (!dir) return;

 while ((d = readdir(dir)))
  {
   if (d->d_name[0] == '.') continue;
   if (!stralloc_copys(tmpname,"tmp/")) break;
   if (!stralloc_cats(tmpname,d->d_name)) break;
   if (!stralloc_0(tmpname)) break;
   if (stat(tmpname->s,&st) == 0)
     if (time > st.st_atime + 129600)
       unlink(tmpname->s);
  }
 closedir(dir);
}

static int append(pq,filenames,subdir,time)
prioq *pq;
stralloc *filenames;
char *subdir;
datetime_sec time;
{
 DIR *dir;
 direntry *d;
 struct prioq_elt pe;
 unsigned int pos;
 struct stat st;

 dir = opendir(subdir);
 if (!dir)
   STRERR_SYS3(-1,maildir_scan_err,"unable to scan $MAILDIR/",subdir,": ")

 while ((d = readdir(dir)))
  {
   if (d->d_name[0] == '.') continue;
   pos = filenames->len;
   if (!stralloc_cats(filenames,subdir)) break;
   if (!stralloc_cats(filenames,"/")) break;
   if (!stralloc_cats(filenames,d->d_name)) break;
   if (!stralloc_0(filenames)) break;
   if (stat(filenames->s + pos,&st) == 0)
     if (st.st_mtime < time) /* don't want to mix up the order */
      {
       pe.dt = st.st_mtime;
       pe.id = pos;
       if (!prioq_insert(pq,&pe)) break;
      }
  }

 closedir(dir);
 if (d) STRERR_SYS3(-1,maildir_scan_err,"unable to read $MAILDIR/",subdir,": ")
 return 0;
}

int maildir_scan(pq,filenames,flagnew,flagcur)
prioq *pq;
stralloc *filenames;
int flagnew;
int flagcur;
{
 struct prioq_elt pe;
 datetime_sec time;
 int r;

 if (!stralloc_copys(filenames,"")) return 0;
 while (prioq_min(pq,&pe)) prioq_delmin(pq);

 time = now();

 if (flagnew) if (append(pq,filenames,"new",time) == -1) return -1;
 if (flagcur) if (append(pq,filenames,"cur",time) == -1) return -1;
 return 0;
}
